doc/acl/Dynamic-ACL-Design.md

```diff
@@ -0,0 +1,198 @@
+                                    "config":{
+                                        "source-ip-address":"192.168.0.2/32",
+                                        "destination-ip-address":"192.168.0.3/32"
+                                    }
+                                },
+                                "dynamic-acl":{
+                                    "config":{
+                                        "ttl":"10"
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+```
+The YANG of ACL_RULE is required to be updated to accept new field `is_dynamic`, `ttl`, `creation_time` and `expiration_time`
+
+Orchagent (actually `aclorch`) won't consume the value of the new flags. So no change is required to `orchagent`.
+#### APP DB
+Existing table `ACL_RULE_TABLE` is used to activate an effective ACL rule. The ACL rule format is the same as in CONFIG DB.
+
+A sample
+```json
+{
+    "APP_ACL_RULE|DATAACL|DYNAMIC_RULE_1":{
+        "DST_IP":"192.168.0.3/32",
+        "ETHER_TYPE":"2048",
+        "PACKET_ACTION":"FORWARD",
+        "PRIORITY":"9999",
+        "SRC_IP":"192.168.0.2/32",
+    }
+}
+```
+#### acl-loader
+
+Update `acl-loader` script to parse new field `ttl`. The entry will be created in `DYNAMIC_ACL_RULE_TABLE` in `CONFIG DB` if `ttl` is present for certain an ACL rule in `acl.json`. Please find more details in workflow diagram.  
+
+#### dynamic_acl_mgrd
+
+A helper script will be added to `swss` container. The checker is started after `orchagent` and check the TTL of dynamic ACL rules every 10 seconds by default. It will walk through all entries in `DYNAMIC_ACL_RULE_TABLE`. New effective ACL rules are added to `APP_ACL_RULE` in `APP DB` and expired ACL rules are deleted both from `CONFIG DB` and `APP DB`.
+
+### Work flow
+#### Add and remove time-based ACL rule
+<p align=center>
+<img src="img/time-based-acl-add-rule.png" alt="Figure 1. Create dynamic ACL rule workflow">
+</p>
+
+#### Refresh TTL of existing ACL rule
+<p align=center>
+<img src="img/time-based-acl-refresh.png" alt="Figure 2. Refresh dynamic ACL rule workflow">
+</p>
+
+### Testing Requirements/Design  
+
+#### Unit Test cases  
+
+1. Enhance unit test for `acl-loader` to verify ACL rule with TTL is accepted; verify `DYNAMIC_ACL_RULE_TABLE` entry is created by CLI in `CONFIG DB`.
+2. Enhance unit test for `acl-loader` to verfiy ACL rule is deleted by CLI in `CONFIG DB`.
+3. Add unit test for `dynamic_acl_mgrd` to verify effective ACL rule is added to `APP DB`; verify expired ACL rule is removed both from `APP DB` and `CONFIG DB`.
+
+#### System Test cases
+
+1. Test case 1 Verify dynamic ACL rule is created as expected
+2. Test case 2 Verify dynamic ACL rule can be refreshed
+3. Test case 3 Verify expired dynamic ACL rule can be removed
+
+## Open questions
+1. memory leak issue detection and validation.
```

Comment:
Do we have more details regarding the mem leak issue?

In other hand, i we can consider follow questions:
1.	When the ACL is triggered, ACL been added to the ASIC by using SAI. The switching performance might be impacted if the ACLs need to be rebuilt on ASIC
2.	The ACL timeout is not able to be handled by SAI/ASIC, software on switch needs to handle all the active Dynamic ACLs
3.	ACLs will increase or decrease the ACL entries dynamically. Too many active ACLs can degrade the switching performance. Need a method to clean up stale rules (rules that can never be actived)

Reply:
Closed